Initial API Structure + Auth Route

xihai01 opened the pull request:
As per discussions in #4934

### What github issue is this PR for, if any?
This PR sets up the initial structure of the API for the app we are building. An auth route for sign in is also added.
Unlike before, the api routes will start with the `/api/v1/*` pattern.
It isn't linked to any issue in this repo - we forked a copy of this repo and are working on that forked copy.

### What changed, and why?
No longer using JWT with devise-jwt gem - instead, opted for a token based authentication approach.
In a nutshell, a token col is created for every row in the user table. The server sends this token in its response body to the client when they sign in. For future requests to server, the client sends this token in its request authorization header.
The reason is because configuring devise to also authenticate (in addition with the current existing web app auth) using devise-jwt was a major headache. I guess it is doable, but would require too much altering existing controllers and files.

I also added active model serializers to help structuring the json data.

Please take a look at the code and give me some feedback! Thank you.

### How is this tested? (please write tests!) 💖💪
I created **api** specs for good/bad login to verify correct headers to ensure correct responses were sent such as token and error msg for bad credentials.
rswag gem used to create api documentations.
Postman was used to verify by hitting the login route for the api.

### Screenshots please :)

#### Postman: Successful login
![Screenshot from 2023-07-11 12-21-25](https://github.com/rubyforgood/casa/assets/86758440/61b30a54-946d-4373-b6ea-037a84ee7e8a)


#### Postman: Unsuccessful login
![Screenshot from 2023-07-11 12-21-52](https://github.com/rubyforgood/casa/assets/86758440/8f8cc6c4-3c47-41bd-8bc1-370b46d85850)


### Feedback please? (optional)
We are very interested in your feedback! Please give us some :) https://forms.gle/1D5ACNgTs2u9gSdh9

xihai01 committed:
implement api structure; add auth routes

xihai01 committed:
lint files

xihai01 commented:
@seanmarcia can I get some feedback please? Thank you!

michaelruhl reviewed (commented):
lgtm, I'd be curious to know if @seanmarcia has any tips though :)

seanmarcia commented:
Hey, sorry for the slow reply. Our large in-person Ruby for Good event is happening this weekend so things have been pretty hectic.

Yes, this is a much better direction and is a great start!

xihai01 committed:
fix some checks like codeclimate and adding extra tests

xihai01 committed:
Merge branch 'main' into api

xihai01 committed:
change file paths of tests so spec checker is satisfied

xihai01 committed:
Merge branch 'api' of github.com:xihai01/casa into api

xihai01 commented:
not to sure why rspec and docker is failing..

littleforest commented on Gemfile:
```diff
@@ -38,13 +38,15 @@ gem "pretender"
 gem "stimulus-rails"
 gem "strong_migrations"
 gem "tzinfo-data", platforms: %i[mingw mswin x64_mingw jruby] # Windows does not include zoneinfo files, so bundle the tzinfo-data gem
+gem "active_model_serializers" # for JSON serialization
```
The active model serializers gem has not been updated in a long time. What do you think about using the [blueprinters](https://github.com/procore-oss/blueprinter) gem instead?

xihai01 replied:
great suggestion! Blueprinters is easier to test as well.

littleforest commented on app/controllers/api/v1/base_controller.rb:
```diff
@@ -0,0 +1,22 @@
+
+  def authenticate_user!
+    token, options = ActionController::HttpAuthentication::Token.token_and_options(request)
+    # return nil unless token && options.is_a?(Hash)
```
Please remove comment

littleforest commented on app/controllers/api/v1/base_controller.rb:
```diff
@@ -0,0 +1,22 @@
+    token, options = ActionController::HttpAuthentication::Token.token_and_options(request)
+    # return nil unless token && options.is_a?(Hash)
+    user = User.find_by(email: options[:email])
+    p "#######"
+    print token
+    p "#######"
```
Please remove STDERR dumps

littleforest commented on app/controllers/api/v1/users/sessions_controller.rb:
```diff
@@ -0,0 +1,28 @@
+  # POST /api/v1/users/sign_in
+  def create
+    load_resource
+    # p @user
```
remove comment

littleforest commented on app/controllers/api/v1/users/sessions_controller.rb:
```diff
@@ -0,0 +1,28 @@
+  end
+
+  def load_resource
+    # print user_params
+    # print user_params
```
Remove comments

littleforest commented on app/controllers/api/v1/users/sessions_controller.rb:
```diff
@@ -0,0 +1,28 @@
+    if !@user&.valid_password?(user_params[:password])
+      @user = nil
+    end
+    # @user = User.find_by(email: user_params[:email])&.valid_password?(user_params[:password])
```
Remove comment

littleforest commented on app/controllers/api/v1/users/sessions_controller.rb:
```diff
@@ -0,0 +1,28 @@
+    # print user_params
+    # print user_params
+    @user = User.find_by(email: user_params[:email])
+    if !@user&.valid_password?(user_params[:password])
```
Suggest:

```suggestion
    unless @user&.valid_password?(user_params[:password])
```

littleforest commented on app/models/concerns/generate_token.rb:
```diff
@@ -0,0 +1,16 @@
+    def generate_hex(column)
+      loop do
+        hex = SecureRandom.hex(32)
+        break hex unless self.class.where(column => hex).any?
```
Rails has some built-in token generating methods. Could you look into using those? https://api.rubyonrails.org/classes/ActiveRecord/SecureToken/ClassMethods.html

littleforest commented on config/environments/development.rb:
```diff
@@ -77,4 +77,7 @@
 
   # Uncomment if you wish to allow Action Cable access from any origin.
   # config.action_cable.disable_request_forgery_protection = true
+  config.hosts << ".ngrok-free.app"
```
How about using an environment variable for this so it doesn't need to be changed across developer environments:

```suggestion
  config.hosts << ENV["DEV_HOSTS"]
```

xihai01 replied:
@littleforest where do we configure the environment variables? I see we are using the dotenv-rails gem but I can't find any .env file that stores env.

xihai01 replied:
can I instead store it in the credentials/development.yml.enc file?

littleforest replied:
@xihai01 you can just create a `.env` file in the root of the app and add the `DEV_HOSTS` environment variable. Just be sure to restart the server once you create and add things to this file.

littleforest commented on config/environments/development.rb:
```diff
@@ -77,4 +77,7 @@
   # config.action_cable.disable_request_forgery_protection = true
+  config.hosts << ".ngrok-free.app"
+
+  config.assets.digest = false
```
Is there a reason why this is getting added?

xihai01 replied:
deleted. I think this was added before when I was trying to configure devise to work with jwt..

littleforest commented on db/migrate/20230710025852_add_token_to_users.rb:
```diff
@@ -0,0 +1,11 @@
+class AddTokenToUsers < ActiveRecord::Migration[7.0]
+  def up
+    add_column :users, :token, :string
+    User.find_each { |user| user.save! }
```
Modifying the user records should be done in an [after party task](https://github.com/theSteveMitchell/after_party)

xihai01 replied:
Why can't you just add a col to the user table with a rails migration instead?

littleforest commented on db/migrate/20230710025852_add_token_to_users.rb:
```diff
@@ -0,0 +1,11 @@
+  def up
+    add_column :users, :token, :string
+    User.find_each { |user| user.save! }
+    # change_column_null :users, :token, false, 1
```
Remove comment

littleforest commented on db/schema.rb:
```diff
@@ -362,6 +362,12 @@
+    t.string "jti", null: false
+    t.datetime "exp", null: false
+    t.index ["jti"], name: "index_jwt_denylist_on_jti"
+  end
```
What is this being used for?

xihai01 replied:
This table was used in the past to store "used" jwt's when devise-jwt gem was used for authentication. Not using devise-jwt anymore, so we can remove this table.

littleforest commented on spec/controllers/api/v1/users/sessions_controller_spec.rb:
```diff
@@ -0,0 +1,26 @@
+
+  it "should handle correct sign in" do
+    post "/api/v1/users/sign_in", {email: volunteer.email, password: volunteer.password}
+    # print last_response.headers
+    # print last_response.body
+    # expect(last_response.headers).to have_key "Authorization"
+    # expect(last_response.headers["Authorization"]).to be_starts_with("Bearer")
```
Remove comments

littleforest commented on spec/controllers/api/v1/users/sessions_controller_spec.rb:
Please write a request spec instead of a controller spec.

xihai01 replied:
when would it be appropriate to write controller specs?

littleforest replied:
Controller specs for rspec are discouraged. Request specs should be used instead.

littleforest commented on spec/controllers/api/v1/base_controller_spec.rb:
Please write a request spec instead of a controller spec. You can see examples in the `spec/requests` directory.

xihai01 replied:
done

littleforest commented on spec/controllers/api/v1/base_controller_spec.rb:
```diff
@@ -0,0 +1,33 @@
+    let(:user) { create(:volunteer) }
+    it "returns http success" do
+      request.headers["Authorization"] = "Token token=#{user.token}, email=#{user.email}"
+      # print request.headers["Authorization"]
```
Remove comment

littleforest commented on spec/controllers/api/v1/base_controller_spec.rb:
```diff
@@ -0,0 +1,33 @@
+      expect(response.body).to eq({message: "Successfully autenticated"}.to_json)
+    end
+    # assert authenticate_user! works with invalid email
+    it "returns http unauthorized" do
```
Remove comment and modify spec description:

```suggestion
    it "returns http unauthorized if invalid email" do
```

xihai01 replied:
done

littleforest commented on spec/controllers/api/v1/base_controller_spec.rb:
```diff
@@ -0,0 +1,33 @@
+  # assert authenticate_user! works
+  describe "GET #index" do
+    let(:user) { create(:volunteer) }
+    it "returns http success" do
```
Remove comment and modify spec description:

```suggestion
    it "returns http success when valid credentials" do
```

xihai01 replied:
done

littleforest reviewed (commented):
Thanks for getting a start on this. We will also need API documentation. I personally like to use the [rswag gem](https://github.com/rswag/rswag) which has you create request specs, and from there it will generate Swagger API documentation.

xihai01 committed:
add rswag gem for api documentation

xihai01 committed:
update gemfile; add blueprinter and remove active model serializer

xihai01 committed:
remove active model serializer files

xihai01 committed:
update files with blueprinter and create request specs

xihai01 committed:
remove comments and add lint files

xihai01 committed:
Merge branch 'main' into api

xihai01 commented:
> @xihai01 thanks for your changes. There are a lot of unaddressed comments and requests in the PR from the previous review. Also, could you please merge or rebase and fix the merge conflict?

ah, my bad. It seems I missed them. I'll work on it right away!

xihai01 committed:
resolve pr changes

xihai01 committed:
Merge branch 'main' into api

xihai01 commented:
@littleforest I've addressed your suggestions and changes!

xihai01 commented:
@littleforest any thoughts? Is it good to merge?

littleforest commented on spec/requests/api/v1/base_spec.rb:
```diff
@@ -0,0 +1,33 @@
+    Rails.application.routes.draw do
+      get "/index", to: "base#index"
+    end
+  end
```
This is okay for now, but let's remove once we have a route we can actually test authentication against.

xihai01 replied:
alright.

littleforest reviewed (approved):
@xihai01 thank you for addressing all of those changes. Sorry for the delay in review.

littleforest commented:
@xihai01 looks like there are a number of merge conflicts now (sorry again for the delay). Can you merge in main and address the conflicts?

xihai01 committed:
Merge branch 'main' into api
